### PR TITLE
ParallelRunner for launching analyses on Barnacle

### PR DESCRIPTION
**Pull request by patrickimran**
This PR adds a ParallelRunner for launching analyses on Barnacle. 
We factor out the preprocessing steps from run_analysis() into run_preprocessing(). In ParallelRunner, it runs the preprocessing for the analysis and then generates the files and file structure needed for mlab to run these jobs on Barnacle, then it submits the job array for that analysis. 

Currently the job submission parameters (chunk size, memory, etc) are hardcoded into the ParallelRunner and not part of AnalysisConfig. 
Currently it only generates results and places them in a folder in `./datasets`, and does not parse through results artifacts nor make any figures/summaries.

### CODE REVIEW

File: imsms_analysis/analysis_runner.py [hunk 1]
```diff
@@ -73,6 +77,105 @@ def run(self, factory):
...
+                    preparation=preparation,
+                    target=target_name,
+                    algorithm=algorithm,
+                    repeats=3,  # num CV repeats
+                    base_dir=base_dir,  # Directory with mlab structure containing datasets
+                    ppn=4,  # processors per node
+                    memory=32,  # memory in GB
+                    wall=50,  # walltime  in hours
+                    chunk_size=chunk_size,  # num parameter sets to run per job
+                    randomize=True,  # randomly shuffle order of parameter set
```
Comment (dhakim87) [inline imsms_analysis/analysis_runner.py:168]: Permuting order of features is very similar to changing random seed in RandomForest right?  I suspect that randomization here will change accuracy by around 2-3% in either direction, so if there aren't enough runs to average that away, it may disguise the effects of other factors.  I think I'd prefer everything to be fully deterministic by default at the moment, setting randomize to False, even though I believe this could mask effects of factors that change the feature set.

File: imsms_analysis/analysis_runner.py [hunk 2]
```diff
@@ -73,6 +77,105 @@ def run(self, factory):
...
+                    force=False,  # force overwrite of existing results
+                    dry=False,  # dry runs
+                    dataset_path=table_fp,
+                    metadata_path=metadata_fp,
+                )
+                
+                cmd = ["qsub", "-t", f"{start}-{end}%{n_concurrent_jobs}", script_fp]
+                subprocess.run(cmd)
```
Comment (dhakim87) [inline imsms_analysis/analysis_runner.py:176]: What is the -t?  Is f"{start}-{end}%{n_concurrent_jobs}" supposed to be the name of the job?  If so, can we append the name in the config?  It looks like all configs will have the same arguments passed in, just different script_fp values.

File: setup.py [hunk 3]
```diff
@@ -2,10 +2,17 @@
 
 setup(
     name="imsms_analysis",
+    package_data={'imsms_analysis':
+        [
+            "dataset/*"
```
Comment (dhakim87) [inline setup.py:7]: Is this recursive?  I believe most of the files are in dataset/biom/* for example.  

This may dramatically increase installation time (I think one of the biom files I have locally is 12 GB).  Is it required to copy data around?  I would prefer to think of data as being separate from the code in this repository at least by design, even though for convenience we may include some right now.  May be out of scope for this PR, but I could add a separate dataset_location field to the config that would let us avoid this copy on installation right?

File: imsms_analysis/analysis_runner.py [hunk 4]
```diff
@@ -73,6 +77,105 @@ def run(self, factory):
...
+                (
+                    script_fp,
+                    params_fp,
+                    run_info_fp,
+                ) = orchestrate_hyperparameter_search(
+                    dataset=dataset,
+                    preparation=preparation,
+                    target=target_name,
+                    algorithm=algorithm,
+                    repeats=3,  # num CV repeats
```
Comment (dhakim87) [inline imsms_analysis/analysis_runner.py:162]: I thought we used to use 5?  Or am I thinking of the relative sizes of training and validation sets?  Is this a change from what we use in SerialRunner or is it exactly the same?

File: imsms_analysis/analysis_runner.py (hunk 1 above)
Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:168]: Randomize here means to do a randomized grid search of the hyperparameters - doesn't affect features at all. So if you run the first 50 parameter sets, you are effectively running a randomized grid search of size 50. 
Recommend keeping this on unless you want to run jobs for an algorithm's entire parameter space (RandomForest's entire parameter space is >10,000 parameter sets but other algorithms are much smaller, <1000).

Otherwise, the parameter sets you explore are ordered according to what's returned by SKLearn's ParameterGrid which is basically just itertools.product, and if you run jobs for the first few parameter sets it wouldn't be a random sample of the hyperparameter space.

File: imsms_analysis/analysis_runner.py (hunk 2 above)
Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:176]: -t is a flag to qsub telling it this is an array job, and it accepts arguments in the form {start}-{end}%{n_concurrent_jobs} so `qsub -t 1:50%10 jobscript.sh` will launch an array job with indices 1...50 inclusive (first 50 chunks, which  each have `chunk_size` parameter sets), and only allow up to 10 of those jobs to run at once.

File: imsms_analysis/analysis_runner.py (hunk 1 above)
Comment (dhakim87) [reply imsms_analysis/analysis_runner.py:168]: Ahhhhh, understood.  Thought this referred to the columns of the input table.  Should be fine for matching SerialRunner, which only uses one set of parameters then.  Is the randomized grid search seeded in any way?  I would still like to preserve determinism if I run a job with 10000 possible parameter set values but only want to choose 100 of them.  Ideally it would always choose the same 100 if I ran the task three times without specifying a different seed value.

File: imsms_analysis/analysis_runner.py (hunk 2 above)
Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:176]: The job name is given by the  template in `q2_mlab/templates/array_job_template.sh `and is made of the dataset, target, 'level' (which is normally 16S or MG, but here we use the analysis name as 'level'), and algorithm.

**Review by dhakim87 — COMMENTED** [review]
Looks pretty good, I had only a couple of minor comments.  Will merge after I give it a full try on Barnacle.  Thanks for this!

File: imsms_analysis/analysis_runner.py (hunk 4 above)
Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:162]: We do a 5-fold cross validation (20-80 split), which might be what you're thinking of. Repeats just means we do this 3 times with  different partitions, for a total of 15 folds.

File: setup.py (hunk 3 above)
Comment (patrickimran) [reply setup.py:7]: I think it is recursive - haven't noticed a slowdown with install times though. Actually this is probably unnecessary so long as you always run `python whateveranalysis.py` from within `imsms_analysis/analyses/` and you have the right relative path to the data.

File: imsms_analysis/analysis_runner.py (hunk 1 above)
Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:168]: Yup, I [seeded it](https://github.com/knightlab-analyses/regression-benchmarking/blob/48cebde806e93be5f5a9ee382e92d14a9b49a6de/q2_mlab/orchestrator.py#L138) for the exact same reason

File: imsms_analysis/analysis_runner.py (hunk 2 above)
Comment (dhakim87) [reply imsms_analysis/analysis_runner.py:176]: Ahh, so if I say start=1, end=10, it will run 10 jobs within the array job, each of which will randomly choose chunk_size parameter sets to try out?   

It looks like this will start one job array per config then, rather than one large job array containing all configs.  That may make it harder to figure out when the job is complete and whether or not all configs processed correctly.

Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:176]: > Ahh, so if I say start=1, end=10, it will run 10 jobs within the array job, each of which will randomly choose chunk_size parameter sets to try out?
yep, more exactly it's taking the first 10 chunks from the randomly-ordered parameter list. If chunk_size=20, the job 1 will be the first 20 parameter sets, job 2 the next 20, etc. 

And also yup, one job array per config. Each job array is a hyperparameter search for that config, split into chunks over several jobs.

Comment (patrickimran) [reply imsms_analysis/analysis_runner.py:176]: > That may make it harder to figure out when the job is complete and whether or not all configs processed correctly.

yeah...I'm working on something for that.

File: imsms_analysis/analysis_runner.py [hunk 5]
```diff
@@ -73,6 +77,105 @@ def run(self, factory):
...
+                    wall=50,  # walltime  in hours
+                    chunk_size=chunk_size,  # num parameter sets to run per job
+                    randomize=True,  # randomly shuffle order of parameter set
+                    force=False,  # force overwrite of existing results
+                    dry=False,  # dry runs
+                    dataset_path=table_fp,
+                    metadata_path=metadata_fp,
+                )
+                
+                cmd = ["qsub", "-t", f"{start}-{end}%{n_concurrent_jobs}", script_fp]
```
Comment (patrickimran) [inline imsms_analysis/analysis_runner.py:175]: Adding a note here to remind myself to add logic that caps the value of `end` at `num_parameter_sets/chunk_size` so we're not submitting jobs for parameter sets that don't exist. I usually do this when I submit jobs for algorithms with a small number of parameter sets (e.g. LinearSVR/SVC has only 16)

**Review by dhakim87 — APPROVED** [review]
